### PR TITLE
Add initial implementation of distributed locks

### PR DESCRIPTION
**Pull request by maraino**
# Description

Add initial implementation of a distributed Lock and TryLock.

### CODE REVIEW

File: recipes/utils.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	if seq == lowestSeq {
+		return lowestNode, nil
+	}
+
+	return lowestNode, ErrLockFound
+}
+
+// timeBasedCleaner deletes those znodes in the base path older than t.
+func timeBasedCleaner(client *ezk.Client, base string, t time.Duration) error {
+	now := unixMilli()
```
Comment (glycerine) [inline recipes/utils.go:92]: this seems to be returning milliseconds, but the rest of the timeout computation is done in seconds. Update: 2 out of 3 of these times are milliseconds

File: recipes/utils.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+	// Read all the childrens
+	children, _, err := client.Children(base)
+	if err != nil {
+		return err
+	}
+
+	// Iterate to all the childrents reading the creation time
+	seconds := int64(t / time.Second)
```
Comment (glycerine) [inline recipes/utils.go]: actually Ctime is in milliseconds too, so this should probably be millis := t / time.Millisecond

File: recipes/utils.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	for i := range children {
+		node := join(base, children[i])
+		ok, stat, err := client.Exists(node)
+		if err != nil {
+			return err
+		}
+
+		// Delete locks older than SchedulerLockTime
+		if ok && stat.Ctime+seconds < now {
```
Comment (glycerine) [inline recipes/utils.go]: then this becomes stat.Ctime + millis

File: recipes/lock.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	return &Lock{
+		Path:   path,
+		client: client,
+		acl:    acl,
+	}
+}
+
+// WithCleaner adds a cleaner to the lock, this will remove all the files
+// in the Lock.Path older than t.
+func (l *Lock) WithCleaner(t time.Duration) {
```
Comment (glycerine) [inline recipes/lock.go:35]: Could we add a comment for the big picture here? To play devil's advocate, why use a lock with cleaner when the lock is already ephemeral and will disappear when the last client detaches?

File: recipes/utils.go (hunk 2 above)
Comment (maraino) [inline recipes/utils.go]: I had it properly implemented before, but with a comment saying that it was seconds, when I reimplemented for ezk, I did not realized it was actually millisecond. 
Fixing it

**Comment by glycerine** [comment]
lgtm

**Pull request merged**